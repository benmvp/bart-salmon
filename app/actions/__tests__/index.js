// @flow
import {setOrigin, setDestination, setNumSalmonRoutes, getSalmonInfo} from '../'

jest.mock('../../api')

describe('setOrigin', () => {
    it('returns SET_ORIGIN action w/ specified station', () => {
        let name = 'WDUB'
        let action = setOrigin(name)

        expect(action).toEqual({
            type: 'SET_ORIGIN',
            payload: {name}
        })
    })
})

describe('setDestination', () => {
    it('returns SET_DESTINATION action w/ specified station', () => {
        let name = 'NCON'
        let action = setDestination(name)

        expect(action).toEqual({
            type: 'SET_DESTINATION',
            payload: {name}
        })
    })
})

describe('setNumSalmonRoutes', () => {
    it('returns SET_NUM_SALMON_ROUTES action w/ specified amount', () => {
        let numRoutes = 19
        let action = setNumSalmonRoutes(numRoutes)

        expect(action).toEqual({
            type: 'SET_NUM_SALMON_ROUTES',
            payload: {numRoutes}
        })
    })
})

describe('getSalmonInfo', () => {
    it('dispatches FETCH_SALMON_INFO action', async () => {
        let mockDispatch = jest.fn()
        let asyncAction = getSalmonInfo()
        let getState = () => ({
            origin: '24TH',
            destination: 'RICH',
            numSalmonRoutes: 7,
            isFetchingSalmonRoutes: false,
            salmonRoutes: [],
            arrivals: []
        })

        await asyncAction(mockDispatch, getState)

        expect(mockDispatch).toHaveBeenCalledTimes(2)
        expect(mockDispatch).toHaveBeenCalledWith({type: 'FETCH_SALMON_INFO'})
    })

    it('dispatches RECEIVE_SALMON_INFO action', async () => {
        let mockDispatch = jest.fn()
        let asyncAction = getSalmonInfo()
        let getState = () => ({
            origin: 'DELN',
            destination: 'EMBR',
            numSalmonRoutes: 3,
            isFetchingSalmonRoutes: false,
            salmonRoutes: [],
            arrivals: []
        })

        await asyncAction(mockDispatch, getState)

        expect(mockDispatch).toHaveBeenCalledTimes(2)

        let actualReceiveActionType = mockDispatch.mock.calls[1][0]

        expect(actualReceiveActionType).toMatchSnapshot()
    })

    it('dispatches ERROR_SALMON_INFO action', async () => {
        let mockDispatch = jest.fn()
        let asyncAction = getSalmonInfo()
        let getState = () => ({
            origin: '24TH',
            destination: '24TH',
            numSalmonRoutes: 7,
            isFetchingSalmonRoutes: false,
            salmonRoutes: [],
            arrivals: []
        })

        await asyncAction(mockDispatch, getState)

        expect(mockDispatch).toHaveBeenCalledTimes(2)

        let actualErrorActionType = mockDispatch.mock.calls[1][0]

        expect(actualErrorActionType).toEqual({
            type: 'ERROR_SALMON_INFO',
            error: expect.any(Error)
        })
    })
})
