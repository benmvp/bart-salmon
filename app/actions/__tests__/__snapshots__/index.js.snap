exports[`getSalmonInfo dispatches RECEIVE_SALMON_INFO action 1`] = `
Object {
  "payload": Object {
    "arrivals": Array [
      Object {
        "abbreviation": "MLBR",
        "bikeflag": 1,
        "color": "RED",
        "destination": "Millbrae",
        "direction": "South",
        "hexcolor": "#ff0000",
        "length": 9,
        "limited": 0,
        "minutes": 12,
        "platform": 2,
      },
      Object {
        "abbreviation": "MLBR",
        "bikeflag": 1,
        "color": "RED",
        "destination": "Millbrae",
        "direction": "South",
        "hexcolor": "#ff0000",
        "length": 9,
        "limited": 0,
        "minutes": 27,
        "platform": 2,
      },
      Object {
        "abbreviation": "MLBR",
        "bikeflag": 1,
        "color": "RED",
        "destination": "Millbrae",
        "direction": "South",
        "hexcolor": "#ff0000",
        "length": 9,
        "limited": 0,
        "minutes": 42,
        "platform": 2,
      },
    ],
    "routes": Array [
      Object {
        "backwardsRideTime": 4,
        "backwardsRouteId": "ROUTE 8",
        "backwardsStation": "RICH",
        "backwardsTrain": Object {
          "abbreviation": "RICH",
          "bikeflag": 1,
          "color": "RED",
          "destination": "Richmond",
          "direction": "North",
          "hexcolor": "#ff0000",
          "length": 9,
          "limited": 0,
          "minutes": 2,
          "platform": 1,
        },
        "backwardsWaitTime": 15,
        "returnRideTime": 4,
        "returnRouteId": "ROUTE 7",
        "returnTrain": Object {
          "abbreviation": "MLBR",
          "bikeflag": 1,
          "color": "RED",
          "destination": "Millbrae",
          "direction": "South",
          "hexcolor": "#ff0000",
          "length": 9,
          "limited": 0,
          "minutes": 21,
          "platform": 2,
        },
        "waitTime": 2,
      },
      Object {
        "backwardsRideTime": 5,
        "backwardsRouteId": "ROUTE 3",
        "backwardsStation": "RICH",
        "backwardsTrain": Object {
          "abbreviation": "RICH",
          "bikeflag": 1,
          "color": "ORANGE",
          "destination": "Richmond",
          "direction": "North",
          "hexcolor": "#ff9933",
          "length": 6,
          "limited": 0,
          "minutes": 9,
          "platform": 1,
        },
        "backwardsWaitTime": 7,
        "returnRideTime": 4,
        "returnRouteId": "ROUTE 7",
        "returnTrain": Object {
          "abbreviation": "MLBR",
          "bikeflag": 1,
          "color": "RED",
          "destination": "Millbrae",
          "direction": "South",
          "hexcolor": "#ff0000",
          "length": 9,
          "limited": 0,
          "minutes": 21,
          "platform": 2,
        },
        "waitTime": 9,
      },
      Object {
        "backwardsRideTime": 4,
        "backwardsRouteId": "ROUTE 8",
        "backwardsStation": "RICH",
        "backwardsTrain": Object {
          "abbreviation": "RICH",
          "bikeflag": 1,
          "color": "RED",
          "destination": "Richmond",
          "direction": "North",
          "hexcolor": "#ff0000",
          "length": 9,
          "limited": 0,
          "minutes": 2,
          "platform": 1,
        },
        "backwardsWaitTime": 30,
        "returnRideTime": 4,
        "returnRouteId": "ROUTE 7",
        "returnTrain": Object {
          "abbreviation": "MLBR",
          "bikeflag": 1,
          "color": "RED",
          "destination": "Millbrae",
          "direction": "South",
          "hexcolor": "#ff0000",
          "length": 9,
          "limited": 0,
          "minutes": 36,
          "platform": 2,
        },
        "waitTime": 2,
      },
    ],
  },
  "type": "RECEIVE_SALMON_INFO",
}
`;
